### ISSUE TITLE
JSON dump lists 'variable' parameter as required for Variables/InitialCondition blocks

### ISSUE THREAD

**Issue by brandonlangley** [idaholab/moose#14232]
## Bug Description
There are many valid MOOSE test inputs with `Variables/InitialCondition/<FunctionIC>` blocks that do not have a `variable` parameter present, e.g.,

```
​[Variables]
  [./c]
    [./InitialCondition]
      type = FunctionIC
      function=(sin(4+x*8-y)*cos(y*10+1.5*x))^2
    [../]
  [../]
[]
```

There is no `variable` parameter present, also there is no `variable` parameter present in the `[GlobalParams]` block.

However, the JSON input definition dump states that this is required, i.e.,

```
/blocks/Variables/star/subblocks/InitialCondition/types/FunctionIC/parameters/variable/required/value(true)
```

Since this happens in many legal inputs, it appears that the JSON input definition dump is incorrect and the `variable` parameter should not be listed as required.

## Steps to Reproduce
Examine the JSON input definition dump produced from running MOOSE with the `--json` command line parameter.

## Impact
The NEAMS Workbench is translating the JSON input definition dump to a schema which is used for input validation and autocomplete assistance across all MOOSE based applications.

Every input with a `Variables/InitialCondition/<FunctionIC>` block that does not contain a `variable` parameter is being reported as invalid.

**Comment by brandonlangley** [comment]
@permcody -

Just FYI, this is one of the two issues that you recommended via our email correspondence that I open. (2/2)

**Comment by rwcarlsen** [comment]
The information is "required", but it is not necessarily required to come _directly_ from syntax in the input file.  In this case, it is implied by the nesting structure - and the moose initialization process handles this implicit `variable=` case by setting the variable parameter after the input file is parsed using this nesting info.  I'm not sure how to best handle this in a way that solves your problem.  Moose internally ensures we always get that param set, but it's not immediately obvious how to make this distinction clear for input file parsing.   What you really need for input file parsing/validation is not `is_param_required` but is actually `is_param_required_directly_in_input_file`.  Moose doesn't really have a way to track/specify the latter.  I'll ponder this a bit.

**Comment by brandonlangley** [comment]
@rwcarlsen -

Thank you for looking into this. Just to make sure I understand - how would this affect the `--json` dump?

Currently, the `SONDefinitionFormatter` just uses the information in the `JsonSyntaxTree` to build the Workbench schema.

So, with an input component with JSON such as this:

```
"variable": {
    "basic_type": "String",
    "cpp_type": "VariableName",
    "deprecated": false,
    "description": "The variable this initial condition is supposed to provide values for.",
    "group_name": "",
    "name": "variable",
    "options": "",
    "out_of_range_allowed": false,
    "required": true
}
```

the `SONDefinitionFormatter` translates the `"required": true` into meaning **required in the input**.

I don't exactly understand the fallout of the suggested change.

Would this add new info into the `JsonSyntaxTree` and I would then need to modify the `SONDefinitionFormatter` to use?

Or, does this suggested change simply flip the `"required": true` to `"required": false` in the places where it is not required in the input file so that the `SONDefinitionFormatter` does not need to be modified and it automatically rolls out?

**Comment by rwcarlsen** [comment]
For actions that generate moose objects, there would be an optional _action_ parameter specifying what, if any, parameters the action sets for the objects created by it (i.e. objects blocks nested underneath the action block's syntax).  So for `[Variable][foo][InitialCondition]type=...` the `Variable/InitialCondition` parameters would have this `_object_params_set_by_action` param listing every parameter that the action sets for the object(s) described underneath it.

**Comment by brandonlangley** [comment]
Okay, so the JSON would have something like this:

```
foo:
 - _object_params_set_by_action=[ foo02 foo04 ]
 - foo01
 - foo02
 - foo03
 - foo04
```

indicating that `foo/foo02` and `foo/foo04` are never "required" in the input (regardless of what the `"required"` field says)?

And for `foo/foo01` and `foo/foo03`, we would just continue to look at the `"required"` field to determine this since they aren't set by `foo`?

**Comment by rwcarlsen** [comment]
The action's parameters (so the superblock of the blocks that have the
required param) would have the _object_params_set_by_action param.

**Comment by brandonlangley** [comment]
@rwcarlsen -

Ahh, okay. I think that make's sense. Thanks!

So, in the `SONDefinitionFormatter` logic, I should just be able to propagate that information down to turn off the "requiredness" of certain parameters in certain contexts.

Will the `_object_params_set_by_action` be an array of parameter names that are set by the object?

This sounds like a good path forward.

I'll wait until this is merged to open my PR for the minor https://github.com/idaholab/moose/issues/16165 change so I can include this `SONDefinitionFormatter` update on that PR too.

**Comment by rwcarlsen** [comment]
Yes - it will be a list/vector/array of param names.  I saw you submitted outputs/file_base and a variables/initialcondition cases for this issue.  Are there any other cases you know about that need to be "fixed"?  The updates are in #16178 so far.
